test/680_proxy_hostname_derivation_test.sh

```diff
@@ -6,23 +6,58 @@ C1=10.2.0.78
 C2=10.2.0.34
 C3=10.2.0.57
 CNAME1=qiuds71y827hdi-seeone-1io9qd9i0wd
-NAME1=seeone.weave.local
+HNAME1=seeone
+FQDN1=$HNAME1.weave.local
 CNAME2=124DJKSNK812-seetwo-128hbaJ881
-NAME2=seetwo.weave.local
+HNAME2=seetwo
+FQDN2=$HNAME2.weave.local
 CNAME3=doesnotmatchpattern
-NAME3=doesnotmatchpattern.weave.local
+HNAME3=doesnotmatchpattern
+FQDN3=$HNAME3.weave.local
 
 start_suite "Hostname derivation through container name substitutions"
 
 weave_on $HOST1 launch-router
 weave_on $HOST1 launch-proxy --hostname-match '^[^-]+-(?P<appname>[^-]*)-[^-]+$' --hostname-replacement '$appname'
 
-proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C1/24 --name=$CNAME1
-proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C2/24 --name=$CNAME2
-proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C3/24 --name=$CNAME3
+CID1=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C1/24 --name=$CNAME1)
+CID2=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C2/24 --name=$CNAME2)
+CID3=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C3/24 --name=$CNAME3)
 
-assert_dns_a_record $HOST1 $CNAME1 $NAME2 $C2
-assert_dns_a_record $HOST1 $CNAME2 $NAME3 $C3
-assert_dns_a_record $HOST1 $CNAME3 $NAME1 $C1
+assert_dns_a_record $HOST1 $CID1 $FQDN1 $C1
+assert_dns_a_record $HOST1 $CID2 $FQDN3 $C3
+assert_dns_a_record $HOST1 $CID2 $FQDN2 $C2
+
+end_suite
+
+
+start_suite "Hostname derivation from container labels"
+
+weave_on $HOST1 launch-router
+weave_on $HOST1 launch-proxy --hostname-from-label hostname-label
+
+CID1=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C1/24 --label=hostname-label=$HNAME1)
+CID2=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C2/24 --label=hostname-label=$HNAME2)
+CID3=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C3/24 --label=hostname-label=$HNAME3)
+
+assert_dns_a_record $HOST1 $CID1 $FQDN1 $C1
+assert_dns_a_record $HOST1 $CID2 $FQDN3 $C3
+assert_dns_a_record $HOST1 $CID3 $FQDN2 $C2
+
+end_suite
+
+
+start_suite "Hostname derivation combining container labels and substitutions"
+
+weave_on $HOST1 launch-router
+weave_on $HOST1 launch-proxy --hostname-from-label hostname-label --hostname-match '^[^-]+-(?P<appname>[^-]*)-[^-]+$' --hostname-replacement '$appname'
+
+CID1=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C1/24 --label=hostname-label=$CNAME1)
+CID2=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C2/24 --label=hostname-label=$CNAME2)
+CID3=$(proxy_start_container_with_dns $HOST1 -e WEAVE_CIDR=$C3/24 --label=hostname-label=$CNAME3)
+
+assert_dns_a_record $HOST1 $CID1 $FQDN1 $C1
+assert_dns_a_record $HOST1 $CID2 $FQDN3 $C3
+assert_dns_a_record $HOST1 $CID3 $FQDN2 $C2
```

Comment:
- please introduce some helper(s) to avoid all the repetition
- this should be a single test suite